File: mobile/modules/orca-mobile-web-shell/android/src/main/java/expo/modules/orcamobilewebshell/MobileWebShellGeneration.kt
package expo.modules.orcamobilewebshell

import java.io.File
import org.json.JSONArray
import org.json.JSONObject

private const val MOBILE_WEB_SHELL_MANIFEST_NAME = "manifest.json"
private const val MOBILE_WEB_SHELL_MANIFEST_CONTENT_TYPE = "application/json"
private const val MOBILE_WEB_SHELL_SCHEMA_VERSION = 1
private const val MOBILE_WEB_SHELL_ENTRYPOINT = "index.html"
private const val MOBILE_WEB_SHELL_MAX_ASSETS = 256
private const val MOBILE_WEB_SHELL_MAX_ASSET_PATH_LENGTH = 255
private const val MOBILE_WEB_SHELL_MAX_CONTENT_TYPE_LENGTH = 128

internal data class MobileWebShellAsset(val file: File, val contentType: String)

/**
 * The served surface of one activated generation: a request path to file map, built once from the
 * manifest before anything loads. Serving is a lookup in this map and never a path join at request
 * time, so "not in the manifest" is a refusal by construction rather than by sanitiser.
 *
 * Asset bytes are not re-hashed here. The TypeScript store verified every byte against the manifest
 * before the activating rename, and the directory path is one the app owns and the page can never
 * influence.
 */
internal class MobileWebShellGeneration private constructor(
  val entries: Map<String, MobileWebShellAsset>
) {
  companion object {
    fun load(directoryPath: String): MobileWebShellGeneration? {
      if (!directoryPath.startsWith("/")) return null
      val directory = File(directoryPath)
      val manifest = runCatching {
        File(directory, MOBILE_WEB_SHELL_MANIFEST_NAME).readText(Charsets.UTF_8)
      }.getOrNull() ?: return null
      return make(manifest, directory)
    }

    fun make(manifestJson: String, directory: File): MobileWebShellGeneration? {
      val root = runCatching { JSONObject(manifestJson) }.getOrNull() ?: return null
      // opt, not optInt: optInt coerces the string "1" to 1, and the contract pins a number.
      if (root.opt("schemaVersion") != MOBILE_WEB_SHELL_SCHEMA_VERSION) return null
      if (root.opt("entrypoint") != MOBILE_WEB_SHELL_ENTRYPOINT) return null
      val assets = root.opt("assets")
      if (assets !is JSONArray) return null
      if (assets.length() == 0 || assets.length() > MOBILE_WEB_SHELL_MAX_ASSETS) return null

      val entries = mutableMapOf<String, MobileWebShellAsset>()
      for (index in 0 until assets.length()) {
        val asset = assets.opt(index)
        if (asset !is JSONObject) return null
        val path = asset.opt("path")
        val contentType = asset.opt("contentType")
        if (path !is String || !isServableAssetPath(path)) return null
        if (contentType !is String || !isServableContentType(contentType)) return null
        entries["/$path"] = MobileWebShellAsset(File(directory, path), contentType)
      }
      // Removed, not copied: the document answers at "/" and nowhere else, so the one response that
      // carries the policy header is the only way to reach those bytes.
      val document = entries.remove("/$MOBILE_WEB_SHELL_ENTRYPOINT") ?: return null
      entries["/"] = document
      // The manifest is written last and is not part of the content hash, so it is not in `assets`;
      // the bootstrap page still reads it from its own origin.
      entries["/$MOBILE_WEB_SHELL_MANIFEST_NAME"] = MobileWebShellAsset(
        File(directory, MOBILE_WEB_SHELL_MANIFEST_NAME),
        MOBILE_WEB_SHELL_MANIFEST_CONTENT_TYPE
      )
      return MobileWebShellGeneration(entries)
    }

    /**
     * Re-checked here rather than trusted: the schema that pins this shape is on the other side of
     * a file the native layer cannot see change.
     */
    fun isServableAssetPath(path: String): Boolean {
      if (path.isEmpty() || path.toByteArray(Charsets.UTF_8).size > MOBILE_WEB_SHELL_MAX_ASSET_PATH_LENGTH) {
        return false
      }
      return path.split('/').all { segment ->
        segment.isNotEmpty() &&
          segment != "." &&
          segment != ".." &&
          segment.all { it in 'a'..'z' || it in 'A'..'Z' || it in '0'..'9' || it == '.' || it == '_' || it == '-' }
      }
    }

    /**
     * This value becomes a response header, so it must not be able to carry a second header or a
     * parameter we did not intend. One lowercase type, one optional charset: the manifest
     * contract's only accepted spelling.
     */
    fun isServableContentType(contentType: String): Boolean {
      if (contentType.isEmpty() ||
        contentType.toByteArray(Charsets.UTF_8).size > MOBILE_WEB_SHELL_MAX_CONTENT_TYPE_LENGTH
      ) {
        return false
      }
      var type = contentType
      val separator = contentType.indexOf("; charset=")
      if (separator >= 0) {
        val charset = contentType.substring(separator + "; charset=".length)
        if (charset.isEmpty()) return false
        if (!charset.all { it in 'a'..'z' || it in '0'..'9' || it == '-' }) return false
        type = contentType.substring(0, separator)
      }
      val halves = type.split('/')
      if (halves.size != 2) return false
      return halves.all(::isMimeToken)
    }

    private fun isMimeToken(token: String): Boolean {
      val first = token.firstOrNull() ?: return false
      if (!(first in 'a'..'z' || first in '0'..'9')) return false
      return token.all { it in 'a'..'z' || it in '0'..'9' || it == '.' || it == '+' || it == '-' }
    }
  }
}

/** `WebResourceResponse` takes the mime type and the encoding separately. */
internal fun splitMobileWebShellContentType(contentType: String): Pair<String, String?> {
  val separator = contentType.indexOf("; charset=")
  if (separator < 0) return contentType to null
  return contentType.substring(0, separator) to
    contentType.substring(separator + "; charset=".length)
}

File: mobile/modules/orca-mobile-web-shell/ios/MobileWebShellGeneration.swift
import Foundation

struct MobileWebShellAsset {
  let file: URL
  let contentType: String
}

enum MobileWebShellGenerationError: Error {
  case unreadable
}

/// The served surface of one activated generation: a request path to file map, built once from the
/// manifest before anything loads. Serving is a lookup in this map and never a path join at request
/// time, so "not in the manifest" is a refusal by construction rather than by sanitiser.
///
/// Asset bytes are not re-hashed here. The TypeScript store verified every byte against the
/// manifest before the activating rename, and the directory path is one the app owns and the page
/// can never influence. Framework-free so `swiftc` can check it.
struct MobileWebShellGeneration {
  static let manifestName = "manifest.json"
  static let manifestContentType = "application/json"
  static let schemaVersion = 1
  static let entrypoint = "index.html"
  static let maxAssets = 256
  static let maxAssetPathLength = 255
  static let maxContentTypeLength = 128

  let entries: [String: MobileWebShellAsset]

  static func load(directoryPath: String) throws -> MobileWebShellGeneration {
    guard directoryPath.hasPrefix("/") else { throw MobileWebShellGenerationError.unreadable }
    let directory = URL(fileURLWithPath: directoryPath, isDirectory: true)
    guard
      let data = try? Data(contentsOf: directory.appendingPathComponent(manifestName))
    else { throw MobileWebShellGenerationError.unreadable }
    return try make(manifestData: data, directory: directory)
  }

  static func make(manifestData: Data, directory: URL) throws -> MobileWebShellGeneration {
    let parsed = try? JSONSerialization.jsonObject(with: manifestData)
    guard
      let root = parsed as? [String: Any],
      isPinnedSchemaVersion(root["schemaVersion"]),
      let declaredEntrypoint = root["entrypoint"] as? String,
      declaredEntrypoint == entrypoint,
      let assets = root["assets"] as? [[String: Any]],
      !assets.isEmpty,
      assets.count <= maxAssets
    else { throw MobileWebShellGenerationError.unreadable }

    var entries: [String: MobileWebShellAsset] = [:]
    for asset in assets {
      guard
        let path = asset["path"] as? String,
        isServableAssetPath(path),
        let contentType = asset["contentType"] as? String,
        isServableContentType(contentType)
      else { throw MobileWebShellGenerationError.unreadable }
      entries["/\(path)"] = MobileWebShellAsset(
        file: directory.appendingPathComponent(path, isDirectory: false),
        contentType: contentType
      )
    }
    // Removed, not copied: the document answers at "/" and nowhere else, so the one response that
    // carries the policy header is the only way to reach those bytes.
    guard let document = entries.removeValue(forKey: "/\(entrypoint)") else {
      throw MobileWebShellGenerationError.unreadable
    }
    entries["/"] = document
    // The manifest is written last and is not part of the content hash, so it is not in `assets`;
    // the bootstrap page still reads it from its own origin.
    entries["/\(manifestName)"] = MobileWebShellAsset(
      file: directory.appendingPathComponent(manifestName, isDirectory: false),
      contentType: manifestContentType
    )
    return MobileWebShellGeneration(entries: entries)
  }

  /// `as? Int` is not this check: NSNumber bridges `true` and `1.0` to 1, and the contract pins the
  /// integer 1. JSONSerialization keeps the written form, so the number's own type answers it.
  static func isPinnedSchemaVersion(_ value: Any?) -> Bool {
    guard let number = value as? NSNumber, CFGetTypeID(number) != CFBooleanGetTypeID() else {
      return false
    }
    let numberType = String(cString: number.objCType)
    guard numberType != "d", numberType != "f" else { return false }
    return number.intValue == schemaVersion
  }

  /// Re-checked here rather than trusted: the schema that pins this shape is on the other side of
  /// a file the native layer cannot see change.
  static func isServableAssetPath(_ path: String) -> Bool {
    guard !path.isEmpty, path.utf8.count <= maxAssetPathLength else { return false }
    for segment in path.split(separator: "/", omittingEmptySubsequences: false) {
      guard !segment.isEmpty, segment != ".", segment != ".." else { return false }
      let valid = segment.allSatisfy { character in
        character.isASCII &&
          (character.isLetter || character.isNumber || character == "." || character == "_" ||
            character == "-")
      }
      guard valid else { return false }
    }
    return true
  }

  /// This value becomes a response header, so it must not be able to carry a second header or a
  /// parameter we did not intend. One lowercase type, one optional charset: the manifest
  /// contract's only accepted spelling.
  static func isServableContentType(_ contentType: String) -> Bool {
    guard !contentType.isEmpty, contentType.utf8.count <= maxContentTypeLength else { return false }
    var type = Substring(contentType)
    if let separator = contentType.range(of: "; charset=") {
      let charset = contentType[separator.upperBound...]
      let validCharset = !charset.isEmpty && charset.allSatisfy { character in
        character.isASCII &&
          (("a"..."z").contains(character) || ("0"..."9").contains(character) || character == "-")
      }
      guard validCharset else { return false }
      type = contentType[contentType.startIndex..<separator.lowerBound]
    }
    let halves = type.split(separator: "/", omittingEmptySubsequences: false)
    guard halves.count == 2 else { return false }
    return halves.allSatisfy(isMimeToken)
  }

  private static func isMimeToken(_ token: Substring) -> Bool {
    guard
      let first = token.first,
      first.isASCII,
      ("a"..."z").contains(first) || ("0"..."9").contains(first)
    else { return false }
    return token.allSatisfy { character in
      character.isASCII &&
        (("a"..."z").contains(character) || ("0"..."9").contains(character) ||
          character == "." || character == "+" || character == "-")
    }
  }
}
